src/BSON/DBPointer.c

```diff
@@ -0,0 +1,380 @@
+#include "php_bson.h"
+
+zend_class_entry *php_phongo_dbpointer_ce;
+
+/* Initialize the object and return whether it was successful. An exception will
+ * be thrown on error. */
+static bool php_phongo_dbpointer_init(php_phongo_dbpointer_t *intern, const char *ref, phongo_zpp_char_len ref_len, const char *id, phongo_zpp_char_len id_len TSRMLS_DC) /* {{{ */
+{
+	if (strlen(ref) != ref_len) {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Ref cannot contain null bytes");
+		return false;
+	}
+
+	if (!bson_oid_is_valid(id, id_len)) {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Error parsing ObjectId string: %s", id);
+		return false;
+	}
+
+	intern->ref = estrndup(ref, ref_len);
+	intern->ref_len = ref_len;
+
+	strncpy(intern->id, id, sizeof(intern->id));
+
+	return true;
+} /* }}} */
+
+/* Initialize the object from a HashTable and return whether it was successful.
+ * An exception will be thrown on error. */
+static bool php_phongo_dbpointer_init_from_hash(php_phongo_dbpointer_t *intern, HashTable *props TSRMLS_DC) /* {{{ */
+{
+#if PHP_VERSION_ID >= 70000
+	zval *ref, *id;
+
+	if ((ref = zend_hash_str_find(props, "ref", sizeof("ref")-1)) && Z_TYPE_P(ref) == IS_STRING &&
+	    (id = zend_hash_str_find(props, "id", sizeof("id")-1)) && Z_TYPE_P(id) == IS_STRING) {
+		return php_phongo_dbpointer_init(intern, Z_STRVAL_P(ref), Z_STRLEN_P(ref), Z_STRVAL_P(id), Z_STRLEN_P(id) TSRMLS_CC);
+	}
+#else
+	zval **ref, **id;
+
+	if (zend_hash_find(props, "ref", sizeof("ref"), (vid**) &ref) == SUCCESS && Z_TYPE_PP(ref) == IS_STRING &&
+	    zend_hash_find(props, "id", sizeof("id"), (void**) &id) == SUCCESS && Z_TYPE_PP(id) == IS_STRING) {
+
+		return php_phongo_dbpointer_init(intern, Z_STRVAL_PP(ref), Z_STRLEN_PP(ref), Z_STRVAL_PP(id), Z_STRLEN_PP(id) TSRMLS_CC);
+	}
+#endif
+
+	phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "%s initialization requires \"ref\" and \"id\" string fields", ZSTR_VAL(php_phongo_dbpointer_ce->name));
+	return false;
+} /* }}} */
+
+/* {{{ proto string MongoDB\BSON\DBPointer::__toString()
+   Return the DBPointer's namespace string and ObjectId. */
+static PHP_METHOD(DBPointer, __toString)
+{
+	php_phongo_dbpointer_t *intern;
+	char                   *retval;
+	int                     retval_len;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+
+	intern = Z_DBPOINTER_OBJ_P(getThis());
+
+	retval_len = spprintf(&retval, 0, "[%s/%s]", intern->ref, intern->id);
+	PHONGO_RETVAL_STRINGL(retval, retval_len);
+	efree(retval);
+} /* }}} */
+
+/* {{{ proto array MongoDB\BSON\Symbol::jsonSerialize()
+*/
+static PHP_METHOD(DBPointer, jsonSerialize)
+{
+	php_phongo_dbpointer_t *intern;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+
+	intern = Z_DBPOINTER_OBJ_P(getThis());
+
+	array_init_size(return_value, 2);
+	ADD_ASSOC_STRINGL(return_value, "$ref", intern->ref, intern->ref_len);
+	ADD_ASSOC_STRING(return_value, "$id", intern->id);
```

Comment:
I just realized that the extended JSON spec calls for: `{"$dbPointer": {"$ref": <namespace [3] as a string>, "$id": ObjectId}}` to avoid confusion with DBRef notation. Can we wrap this in an outer array with a `$dbPointer` key?